src/rgw/rgw_rest_client.cc

```diff
@@ -434,11 +434,35 @@ static void add_grants_headers(map<int, string>& grants, RGWEnv& env, map<string
   }
 }
 
+static void replaceAll(string& src_str, const string& old_str, const string& new_str)
+{
+  size_t pos = 0;
+  while ((pos = src_str.find(old_str, pos)) != string::npos) {
+    src_str.replace(pos, old_str.length(), new_str);
+    pos += old_str.length();
+  }
+}
+
 void RGWRESTStreamS3PutObj::send_init(rgw_obj& obj)
 {
+  string resourceStr;
   string resource;
-  url_encode(obj.bucket.name + "/" + obj.get_oid(), resource);
   string new_url = url;
+
+  if (host_style == VirtualStyle){
+    resourceStr = obj.get_oid();
+    new_url = obj.bucket.name + "."  + new_url;
+  }else {
+    resourceStr = obj.bucket.name + "/" + obj.get_oid();
+  }
+
+  url_encode(resourceStr, resource);
+  replaceAll(resource, "%2F", "/");
```

Comment:
I'd rather have another version of url_encode that gets a list of characters that shouldn't be encoded.